decode.go

```diff
@@ -253,6 +253,16 @@ func (de *decoder) putvalue(wiretype int, val reflect.Value,
 
 	// Embedded message
 	case reflect.Struct:
+		bm := reflect.TypeOf((*encoding.BinaryUnmarshaler)(nil)).Elem()
+		if ok := val.Addr().Type().Implements(bm); ok {
+			// If the object support self-decoding, use that.
+			if wiretype != 2 {
```

Comment:
Can't we put that condition already at the top , just after `reflect.Struct:` instead of writing it twice?